include/libsemigroups/adapters.hpp

```diff
@@ -514,9 +571,123 @@ namespace libsemigroups {
+  //!
+  //! This type should be a stateless struct with a single typedef \c type
+  //! representing the type that should be used or returned by a \c Lambda
+  //! struct.
+  //!
+  //! \tparam TElementType the type of the semigroup elements.
+  //!
+  //! The second template parameter exists for SFINAE in overload resolution.
+  //!
+  //! \par Used by KoniecznyTraits.
+  template <typename TElementType, typename = void>
+  struct LambdaValue;
+
+  //! Defined in ``adapters.hpp``.
+  //!
+  //! This type should be a stateless struct with a single typedef \c type
+  //! representing the type that should be used or returned by a \c Rho
+  //! struct.
+  //!
+  //! \tparam TElementType the type of the semigroup elements.
+  //!
+  //! The second template parameter exists for SFINAE in overload resolution.
+  //!
+  //! \par Used by KoniecznyTraits.
+  template <typename TElementType, typename = void>
+  struct RhoValue;
+
+  //! Defined in ``adapters.hpp``.
+  //!
+  //! This type should be a stateless trivially default constructible
+  //! with an operator of signature `void
+  //! operator()(TPointType&, TElementType const&)`, and an operator of
+  //! signature `TPointType operator()(TElementType const&)`.
+  //!
+  //! The first call operator should modify the first argument in-place to
+  //! contain the lambda value of the second argument. The second call operator
+  //! should return the lambda value of the argument. The lambda values should
+  //! be a constant function on any given \f$\mathscr{L}\f$-class of the
+  //! semigroup in question.
+  //!
+  //! \tparam TElementType the type of the semigroup elements.
+  //! \tparam TPointType the type of the lambda points.
+  //!
+  //! The second template parameter exists for SFINAE in overload resolution.
+  //!
+  //! \par Used by KoniecznyTraits.
+  template <typename TElementType, typename TPointType>
+  struct Lambda;
+
+  //! Defined in ``adapters.hpp``.
+  //!
+  //! This type should be a stateless trivially default constructible
+  //! with an operator of signature `void
+  //! operator()(TPointType&, TElementType const&)`, and an operator of
+  //! signature `TPointType operator()(TElementType const&)`.
+  //!
+  //! The first call operator should modify the first argument in-place to
+  //! contain the rho value of the second argument. The second call operator
+  //! should return the rho value of the argument. The rho values should
+  //! be a constant function on any given \f$\mathscr{R}\f$-class of the
+  //! semigroup in question.
+  //!
+  //! \tparam TElementType the type of the semigroup elements.
+  //! \tparam TPointType the type of the rho points.
+  //!
+  //! The second template parameter exists for SFINAE in overload resolution.
+  //!
+  //! \par Used by KoniecznyTraits.
+  template <typename TElementType, typename TPointType>
+  struct Rho;
+
+  //! Defined in ``adapters.hpp``.
+  //!
+  //! Derived classes of this should have a typedef \c type representing the
```

Comment:
"Derived classes" -> "Specializations"?

Reply:
Also: should it be mentioned here that the `type` is void by default, and this declaration is used to indicate that there's no rank state object required?